src/core/lib/event_engine/posix_engine/traced_buffer_list.h

```diff
@@ -0,0 +1,155 @@
+// Copyright 2022 The gRPC Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef GRPC_CORE_LIB_EVENT_ENGINE_POSIX_ENGINE_TRACED_BUFFER_LIST_H
+#define GRPC_CORE_LIB_EVENT_ENGINE_POSIX_ENGINE_TRACED_BUFFER_LIST_H
+
+#include <grpc/support/port_platform.h>
+
+#include "absl/functional/any_invocable.h"
+#include "absl/status/status.h"
+#include "absl/types/optional.h"
+
+#include <grpc/support/time.h>
+
+#include "src/core/lib/event_engine/posix_engine/internal_errqueue.h"
+#include "src/core/lib/iomgr/port.h"
+
+namespace grpc_event_engine {
+namespace posix_engine {
+
+struct ConnectionMetrics {
+  /* Delivery rate in Bytes/s. */
+  absl::optional<uint64_t> delivery_rate;
```

Comment:
Are there any subsets of these values that will always be present or absent in unison? If so, then a single `absl::has_value<SpecificXYZMetrics>(thing)` is going to be nicer than a handful of `absl::has_value<uint64_t>(x)` calls (and maybe save a few cycles).

Are there any that will always be present? If so, they don't need to be optional.

Reply:
Unfortunately I dont think thats true. Some kernel versions may support some values while others dont.

Reply:
I'm wondering instead if a collection of named metrics would be better. You'd then be able to iterate across key-values, and it'll only contain the data that's actually available.

Reply:
I feel that would require storing a string or some other type key and a value for each unique metric which might be expensive.

Reply:
As discussed offline, this is fine as is. Keys could be enum classes to save some space, but I'm happy to not bikeshed this any further.